datafusion/common/src/scalar.rs

```diff
@@ -101,6 +101,8 @@ pub enum ScalarValue {
     IntervalMonthDayNano(Option<i128>),
     /// struct of nested ScalarValue
     Struct(Option<Vec<ScalarValue>>, Box<Vec<Field>>),
+    /// Dictionary type: index type and value
+    Dictionary(Box<DataType>, Box<ScalarValue>),
```

Comment:
The core change is to introduce this variant and the rest of the PR is implementing it as well as removing the partial support for Dictionaries that existed previously